File: app/src/main/java/com/jeketos/associatewith/drawer/DrawerModelImpl.java
package com.jeketos.associatewith.drawer;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.jeketos.associatewith.Point;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */

public class DrawerModelImpl implements DrawerMVP.DrawerModel {

    private static final String MOVE = "move";
    private static final String CHAT = "chat";
    DrawerMVP.DrawerPresenter presenter;
    private DatabaseReference referenceMove;
    private DatabaseReference referenceChat;

    public DrawerModelImpl(DrawerMVP.DrawerPresenter presenter) {
        this.presenter = presenter;
        referenceMove = FirebaseDatabase.getInstance().getReference(MOVE);
        referenceChat = FirebaseDatabase.getInstance().getReference(CHAT);
        referenceChat.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                presenter.chatDataReceived(dataSnapshot);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });
    }

    //FIXME
    @Override
    public int getMovesCount() {
        return 0;
    }

    @Override
    public void sendPoint(int movesCount, Point point) {
        Map<String, Object> map = new HashMap<>();
        map.put(Integer.toString(movesCount), point);
        referenceMove.updateChildren(map);
    }

    @Override
    public void clearData() {
        referenceMove.removeValue();
        referenceChat.removeValue();
        presenter.clearChat();
    }
}

File: app/src/main/java/com/jeketos/associatewith/drawer/DrawerModelImpl.kt
package com.jeketos.associatewith.drawer

import com.google.firebase.database.*
import com.jeketos.associatewith.Point
import java.util.*
import javax.inject.Inject

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */
 class DrawerModelImpl @Inject constructor() : DrawerMVP.DrawerModel {

    val  MOVE = "move"
    val  CHAT = "chat"
    val WORDS = "words"
    val SELECTED_WORD = "selected_word"
    val referenceMove : DatabaseReference
    val referenceChat : DatabaseReference
    val referenceWords : DatabaseReference
    val referenceSelectedWord : DatabaseReference
    val referenceWinner : DatabaseReference
    lateinit var firebaseListener: (DataSnapshot) -> Unit
    @Inject lateinit var presenter : DrawerMVP.DrawerPresenter

    init {
        referenceSelectedWord = FirebaseDatabase.getInstance().getReference(SELECTED_WORD)
        referenceMove = FirebaseDatabase.getInstance().getReference(MOVE)
        referenceChat = FirebaseDatabase.getInstance().getReference(CHAT)
        referenceWords = FirebaseDatabase.getInstance().getReference(WORDS)
        referenceWinner = FirebaseDatabase.getInstance().getReference("winner")
        referenceChat.addValueEventListener(object : ValueEventListener{
            override fun onCancelled(p0: DatabaseError?) {}

            override fun onDataChange(dataSnapshot: DataSnapshot?) {
                presenter.chatDataReceived(dataSnapshot!!)
            }
        })
        referenceWords.addValueEventListener(object : ValueEventListener{
            override fun onCancelled(p0: DatabaseError?) {}

            override fun onDataChange(dataSnapshot: DataSnapshot?) {
                presenter.wordsDataReceived(dataSnapshot!!)
            }
        })
        referenceWinner.addValueEventListener(object : ValueEventListener{
            override fun onCancelled(p0: DatabaseError?) {}

            override fun onDataChange(dataSnapshot: DataSnapshot?) {
                presenter.winnerDataReceived(dataSnapshot!!)
            }
        })
    }

    override fun getMovesCount(): Int{
        return 0
    }

    override fun sendPoint(movesCount: Int, point: Point) {
        val map = HashMap<String,Any>()
        map.put(movesCount.toString(),point)
        referenceMove.updateChildren(map)
    }

    override fun clearData() {
        referenceMove.removeValue()
        referenceChat.removeValue()
        referenceSelectedWord.removeValue()
        referenceWinner.removeValue()
    }

    override fun saveSelectedWord(word: CharSequence) {
        referenceSelectedWord.setValue(word)
    }

    override fun setListener(firebaseListener: (DataSnapshot) -> Unit) {
        this.firebaseListener = firebaseListener
    }
}
